### PR TITLE
[WIP] replace sprintf to string.format for FSharp.Core.UnitTests

### PR DESCRIPTION
**Pull request by FoggyFinder**
Remove the use of sprintf for FSharp.Core.UnitTests

[#4954](https://github.com/Microsoft/visualfsharp/issues/4954)

### CODE REVIEW

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs [hunk 1]
```diff
@@ -15,8 +15,8 @@ let smallerSizeCheck testable = Check.One({ Config.QuickThrowOnFailure with EndS
 
 /// helper function that creates labeled FsCheck properties for equality comparisons
 let consistency name sqs ls arr =
-    (sqs = arr) |@ (sprintf  "Seq.%s = '%A', Array.%s = '%A'" name sqs name arr) .&. 
-    (ls  = arr) |@ (sprintf "List.%s = '%A', Array.%s = '%A'" name ls name arr)
+    (sqs = arr) |@ (String.Format("Seq.{0} = '{1}', Array.{0} = '{2}'", name, sqs, arr)) .&. 
+    (ls  = arr) |@ (String.Format("List.{0} = '{1}', Array.{0} = '{2}'", name, ls, arr))
```
Comment (vasily-kirichenko) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs]: Are you sure it's safe to replace `%A` with `ToString` here?

Comment (FoggyFinder) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs]: @vasily-kirichenko no, I have strong doubt here. Dunno why I decided to change it. I revert these lines.

**Comment by KevinRansom** [comment]
Just out of curiosity, why?

**Comment by charlesroddie** [comment]
@KevinRansom This will allow adjusted tests to run under CoreRT and .Net Native.

Requested by @dsyme : "If you like you can submit a PR to remove the use of sprintf throughout FSHarp.Core.UnitTests except where that's the thing explicitly being tested.  It's useful anyway to reduce the complexity of what's being executed for different tests."

**Comment by dsyme** [comment]
I'm ok with this as a general engineering improvement, since the dependency basis of the individual tests is then much slimmer. This is useful any time anyone wants to take F# to a different .NET implementation or for Fable/JS, WebSharper, Fez and so on.

**Comment by charlesroddie** [comment]
> @zpodlovics some earlier test cases was carefully designed and intentionally avoided this issue by factoring out the printing functionality to the beginning of the test code

Agreed that this would be nicer, and your first code block (`report_failure` etc.) shows a good structure, but I do not think we need #if definitions as you suggest lower down, because we do not need to maintain sprintf. As @dsyme said, moving away from sprintf (except where sprintf is specifically being tested) would 1. be an improvement all round, simplifying and reducing dependencies, and 2. may apply to more compilation toolchains than we know about at the time of writing the tests.

**Comment by zpodlovics** [comment]
@charlesroddie  Also, in several cases conditional compilation will and still required, because corert does not support (or will not) support some features eg.: non-zero indexed arrays, deeply nested generics

**Comment by FoggyFinder** [comment]
Well, I think all occurrences of sprintf which could easily and safely replacing to String.Format were removed.

There are two sprintf:

```fsharp
(sqs = arr) |@ (sprintf  "Seq.%s = '%A', Array.%s = '%A'" name sqs name arr) .&. 
(ls  = arr) |@ (sprintf "List.%s = '%A', Array.%s = '%A'" name ls name arr)
```

Is it better to replace it with String.Format + custom function for formatting or leave unchanged?

What should to do with `printf`/`bprintf` also?

**Comment by charlesroddie** [comment]
`bprintf sb f = sprintf f >>sb.Append`. That is straightforward.
`printf sb = sprintf sb >> Console.Write`, although should probably be `sprintf sb >> print` so `print` can be defined more appropriately if needed.

**Comment by dsyme** [comment]
To me this as a basically harmless engineering improvement but I don't really care that much either way. I've often thought about doing it myself, especially when I've just inserted some bug deep into F# that affects "sprintf" and I want it caught early and obviously. 

The way I see it is this: the FSharp.Core unittests are especially useful during the process of trialling F# on different runtime substrates. Historical examples where we've done this include WebSharper, Fez, Silverlight, PCL, Fable, testing various F# interpreters and dynamic compilers (Unquote etc.), bringing up F# on .NET Core 1.x etc. Just last week I was writing an F# interpreter to run F# code in a LivePlayer on mobile devices and minimized unit tests would have been useful.

So if it gets done, sure, it's useful, I approve, but I don't care much either way.  

@charlesroddie  Since @KevinRansom doesn't like this, perhaps just make a fork of the tests instead?

**Comment by dsyme** [comment]
> We would support an effort to reduce, or better yet eliminate the requirement for reflection during printfn/printf/sprintf. Obviously this is going to be a lot of effort but would yield significant benefits to any future Ahead of Time Compilation or Transpiler strategies that F# might follow.

This makes sense.  If it helps the history is this:

* Once upon a time (approx F# 0.6) we used compiler-generated code to avoid any reflection, so a printf string would become compiler-generated code.  That can work but was at the time buggy, complex, created large numbers of closures and was not particularly fast.  

* In F# 1.x we moved to runtime interpretation, which needs reflection (dynamic instantiations via MakeGenericMethod and MakeGenericType) but no explicit code generation.  This means it works on Xamarin iOS (where a last-resort imlpementation of generic code is always available, but codegen is not permitted) but not on .NET Native and CoreRT (where no last-resort implementation of generic code is available, sadly)

* In F# 3.x we moved to "fast" runtime interpretation, the current implementation (done by @v2m), with the same dependency base.

This all means I basically put the current F# "printf" implementation in the same category as LINQ: unlikely to work on implementations of .NET that don't support arbitrary MakeGenericType/MakeGenericMethod.    (For this reason  I'd also recommend avoiding the use of LINQ when testing primitives - as a habit but not a rule :) )

cheers
don

**Comment by charlesroddie** [comment]
> @dsyme Since @KevinRansom doesn't like this, perhaps just make a fork of the tests instead?
Yes we'll keep working on this fork for use in manual testing of .Net Native, since changes to printfn/printf/sprintf sound like they won't be done in the short term. We'll preserve this WIP PR for visibility. @FoggyFinder

**Comment by KevinRansom** [comment]
It sounds like you have a plan.  I would really prefer, if there was someone to champion looking at how we need to react to platforms without reflection, although that may I suppose have to wait until we fund it, here which will certainly not be any time soon.

You should also submit any tests you care about to the coreRT repo.  I would hope that they would probably take IL assemblies that you generate.  At least that way, anything you determine works would stay working.

**Pull request closed without merging**